### PR TITLE
Add CUDA 11 builds for Windows CI

### LINKED ISSUE #42549: Caffe2 Error: more than one operator "+" matches these operands, windows and Cuda 11
## 🐛 Bug

An error occurs while building PyTorch from the latest master on Win10 with VS2019 and Cuda 11. The error seems to be from Caffe2 (refer to the error quote below). Thanks in advance.

## To Reproduce

On Windows 10 Home:
1. Download and install the latest NVIDIA drivers and Visual Studio 2019 (with Visual C++).
2. Download and install the latest Cuda Toolkit (Cuda 11).
3. Download and install python 3.8
3. Clone the master branch of PyTorch.
4. Download and install the PyTorch dependencies.
5. In CMD: run "set CMAKE_GENERATOR=Visual Studio 16 2019"
6. In the PyTorch repo directory: run "python setup.py install"

The error:
> [1040/1919] Building NVCC (Device) object caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_DilatedMaxPool2d.cu.obj
> FAILED: caffe2/CMakeFiles/torch_cuda.dir/operators/torch_cuda_generated_utility_ops.cu.obj
> ...[cmd command info here]...
> C:/.../pytorch/build\caffe2/proto/caffe2.pb.h(151): error: more than one operator "+" matches these operands:
>             built-in operator "arithmetic + arithmetic"
>             function "c10::operator+(int, c10::BFloat16)"
>             function "c10::operator+(c10::BFloat16, int)"
>             function "c10::operator+(int, c10::Half)"
>             function "c10::operator+(c10::Half, int)"
>             operand types are: const caffe2::ExternalDataProto_SourceType + int

## Expected behavior

The build should complete successfully. I haven't made any modifications to the clone.

## Environment

PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A

OS: Microsoft Windows 10 Home (ver 19041)
GCC version: Could not collect
CMake version: version 3.18.0

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: Cuda 11
GPU models and configuration: GPU 0: GeForce GT 710 (compute capability: 3.5)
Nvidia driver version: 451.67
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.4
[conda] Could not collect

cc @malfet @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @smartcat2010 @mszhanyi

### PR DESCRIPTION
**Pull request by peterjc123**
Stacked on https://github.com/pytorch/pytorch/pull/42410.

### CODE REVIEW

**Review by kurtamohler — APPROVED** [review]

File: aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu
Comment (malfet) [inline aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu:79]: Hmm, your change disables this functionality on Linux with CUDA-10.2, which is not ideal, is it?
```suggestion
#if defined(__CUDACC__) && (CUSPARSE_VERSION >= 11000 || (!defined(_MSC_VER) && CUSPARSE_VERSION >= 10301))
```

File: aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu
Comment (malfet) [inline aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu:12]: Let's keep this enabled for CUDA 10.2 on Linux
```suggestion
#if defined(__CUDACC__) && (CUSPARSE_VERSION >= 11000 || (!defined(_MSC_VER) && CUSPARSE_VERSION >= 10301))
```

File: aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu
Comment (malfet) [inline aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu:16]: How this is related to CUDA-11 upgrade?

**Comment by ezyang** [comment]
Please don't merge without go-ahead from @malfet , cc'ing @malfet and @zpao for CI cost implications

File: cmake/ProtoBufPatch.cmake
Comment (malfet) [inline cmake/ProtoBufPatch.cmake:55]: How this related to CUDA-11 update?

**Review by malfet — CHANGES_REQUESTED** [review]

**Comment by xwang233** [comment]
Part of the cusparse changes (cuda11+windows) were just merged at https://github.com/pytorch/pytorch/pull/42412.

File: aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu
Comment (xwang233) [reply aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu:79]: You can check cusparse details at https://github.com/pytorch/pytorch/pull/42412. On linux cuda 10.2, using new or old API should have no big differences. They are both functional. This is just to make the macro straightforward.

**Review by malfet — APPROVED** [review]
Please rebase, otherwise LGTM (all the comments I have on the diff, other than Protobuf patch coming from https://github.com/pytorch/pytorch/pull/42412)

File: .circleci/cimodel/data/windows_build_definitions.py
Comment (peterjc123) [reply .circleci/cimodel/data/windows_build_definitions.py]: Removed the one for cuda 11.

File: cmake/ProtoBufPatch.cmake
Comment (peterjc123) [reply cmake/ProtoBufPatch.cmake:55]: The error (in comments) will be raised for CUDA 11 builds.

**Comment by peterjc123** [comment]
@malfet Rebased.

**Review by guyang3532 — APPROVED** [review]

**Review by xwang233 — APPROVED** [review]

**Pull request closed without merging**
